### PR TITLE
Unified brace_linter

### LINKED ISSUE #1041: Unify bracing-related linters / parentheses / if-else clauses?
We currently have a variegated list of linters related to bracing and parentheses:

 - `closed_curly_linter()`
 - `else_same_line_linter()`
 - `function_brace_linter()`
 - `function_left_parentheses_linter()`
 - `if_else_match_braces_linter()`
 - `open_curly_linter()`
 - `paren_body_linter()`
 - `paren_brace_linter()`
 - `spaces_inside_linter()`
 - `spaces_left_parentheses_linter()`

Should we consolidate any of these?

### PR DESCRIPTION
**Pull request by AshesITR**
Fixes #1041

I'll do a chain of PRs into this main one to make review easier.
This first step contains the closed_curly_linter, which I ported to XPath while at it.

Tests were largely copied (one new one because the XML parse data for `a({...})` and `a(function() {...})` differs) from `closed_curly_linter()` but made a little more readable.

 - [x] deprecate closed_curly_linter
 - [x] add brace_linter to defaults instead of closed_curly_linter
 - [x] add breaking change to NEWS

### CODE REVIEW

File: R/brace_linter.R [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#' Brace linter
+#'
+#' Perform various style checks related to placement and spacing of curly braces:
+#'
+#'  - Curly braces are on their own line unless they are followed by an `else`.
+#'
+#' @param allow_single_line if `TRUE`, allow an open and closed curly pair on the same line.
+#'
+#' @evalRd rd_tags("brace_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
```
Comment (MichaelChirico) [inline R/brace_linter.R:10]: copy the style guide link over as well:

```
 \cr
#'   <https://style.tidyverse.org/syntax.html#indenting>
```

File: R/brace_linter.R [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    lints <- list()
+
+    xp_cond_closed <- xp_and(c(
+      # matching { is on same line
+      if (isTRUE(allow_single_line)) {
+        "(@line1 != preceding-sibling::OP-LEFT-BRACE/@line1)"
+      },
+      # immediately followed by "," or ")"
```
Comment (MichaelChirico) [inline R/brace_linter.R:25]: `}]` should be OK too right?

```
DT[, {
  do_stuff
}]
```

File: R/brace_linter.R [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      # double curly
+      "not(
+        (@line1 = parent::expr/following-sibling::OP-RIGHT-BRACE/@line1) or
+        (@line1 = preceding-sibling::expr/OP-RIGHT-BRACE/@line1)
+      )"
+    ))
+
+    xp_closed_curly <- glue::glue("//OP-RIGHT-BRACE[
+      { xp_cond_closed } and (
+        (@line1 = preceding-sibling::*/@line2) or
```
Comment (MichaelChirico) [inline R/brace_linter.R:36]: is `*` and not `*[1]` intentional?

**Comment by MichaelChirico** [comment]
Check the new lints -- new logic is missing something, e.g.

```
file=R/closed_curly_linter.R,line=70,col=12,Closing curly-braces should always be on their own line, unless they are followed by an else.
```

File: tests/testthat/test-brace_linter.R [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    "),
+    list(
+      closed_curly_msg,
+      closed_curly_msg
+    ),
+    linter
+  )
+
+  # }) is allowed
+  expect_lint("eval(bquote({...}))", NULL, linter)
```
Comment (MichaelChirico) [inline tests/testthat/test-brace_linter.R:50]: does this test need `{` and `}` to be on separate lines?

File: tests/testthat/test-brace_linter.R [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  # }, is allowed
```
Comment (MichaelChirico) [inline tests/testthat/test-brace_linter.R:52]: should be clear from above but:

add a `}]` test as well

File: R/brace_linter.R (hunk 1 above)
Comment (MichaelChirico) [reply R/brace_linter.R:10]: nvm I see it in #1094

File: tests/testthat/test-brace_linter.R (hunk 4 above)
Comment (MichaelChirico) [reply tests/testthat/test-brace_linter.R:50]: I see that's done in #1096

**Comment by MichaelChirico** [comment]
Great work on the whole chain!

File: R/brace_linter.R (hunk 2 above)
Comment (AshesITR) [reply R/brace_linter.R:25]: Sounds like a worthwhile addition.

File: R/brace_linter.R (hunk 3 above)
Comment (AshesITR) [reply R/brace_linter.R:36]: Shouldn't matter due to monotonicity of line numbers, but `*[1]` might be more efficient.

**Comment by AshesITR** [comment]
> Check the new lints -- new logic is missing something, e.g.
> 
> ```
> file=R/closed_curly_linter.R,line=70,col=12,Closing curly-braces should always be on their own line, unless they are followed by an else.
> ```

That was actually a true positive, complaining about the closing `}` of the if not being on its own.
Fixed the lint here.

**Comment by AshesITR** [comment]
Filing the `c_style_braces` part for follow-up for now, because it would require changes to most lints.

**Review by MichaelChirico — APPROVED** [review]

**Comment by AshesITR** [comment]
Thanks for the quick review. Ready to merge now, as the chain is completely merged.

**Review by MichaelChirico — APPROVED** [review]

**Pull request merged**